web/src/dropdown_widget.js

```diff
@@ -0,0 +1,134 @@
+import {parse_html} from "./ui_util";
+
+/* Sync with max-height set in popovers.css */
+export const DEFAULT_DROPDOWN_MAX_HEIGHT = 200;
+
+export function setup(popover_props, get_options, item_click_callback) {
+    const on_show_callback = popover_props.on_show_callback;
+    // Since we pass `popover_props` to `tippy`, we should delete our custom properties.
+    delete popover_props.on_show_callback;
+
+    tippy.delegate("body", {
+        ...default_popover_props,
+        // New theme defined in popovers.css
+        theme: "popovers",
+        arrow: false,
+        onShow(instance) {
+            instance.setContent(parse_html(render_dropdown_list_container()));
+            const $popper = $(instance.popper);
+            const $dropdown_list_body = $popper.find(".dropdown-list");
+            const $search_input = $popper.find(".dropdown-list-search-input");
+
+            const list_widget = ListWidget.create($dropdown_list_body, get_options(), {
+                name: `${CSS.escape(popover_props.target)}-list-widget`,
+                modifier(item) {
+                    return render_dropdown_list({item});
+                },
+                filter: {
+                    $element: $search_input,
+                    predicate(item, value) {
+                        return item.name.toLowerCase().includes(value);
+                    },
+                },
+                $simplebar_container: $popper.find(".dropdown-list-wrapper"),
+            });
+
+            // Keyboard handler
+            $popper.on("keydown", (e) => {
+                function trigger_element_focus($element) {
+                    e.preventDefault();
+                    e.stopPropagation();
+                    $element.trigger("focus");
+                }
+
+                const $search_input = $popper.find(".dropdown-list-search-input");
+                const list_items = list_widget.get_current_list();
+                if (list_items.length === 0 && !(e.key === "Escape" || e.key === "Enter")) {
+                    // Let the browser handle it.
+                    return;
+                }
+
+                function first_item() {
+                    const first_item = list_items[0];
+                    return $popper.find(`.list-item[data-unique-id="${first_item.unique_id}"]`);
+                }
+
+                function last_item() {
+                    const last_item = list_items.at(-1);
+                    return $popper.find(`.list-item[data-unique-id="${last_item.unique_id}"]`);
+                }
+
+                switch (e.key) {
+                    case "Enter":
+                        if (list_items.length === 0) {
+                            $(e.target).trigger("click");
+                        } else {
+                            // Select first item.
+                            first_item().trigger("click");
+                        }
+                        e.target.click();
```

Comment:
How does this work -- it seems like we're triggering a click twice in some cases?

Reply:
I was able to reproduce buggy behavior with this logic as written that is fixed with the following changes, which I've squashed in.

```
diff --git a/web/src/dropdown_widget.js b/web/src/dropdown_widget.js
index 042a3d8d02..15051d043d 100644
--- a/web/src/dropdown_widget.js
+++ b/web/src/dropdown_widget.js
@@ -51,7 +51,7 @@ export function setup(popover_props, get_options, item_click_callback) {
 
                 const $search_input = $popper.find(".dropdown-list-search-input");
                 const list_items = list_widget.get_current_list();
-                if (list_items.length === 0 && !(e.key === "Escape" || e.key === "Enter")) {
+                if (list_items.length === 0 && !(e.key === "Escape")) {
                     // Let the browser handle it.
                     return;
                 }
@@ -68,13 +68,12 @@ export function setup(popover_props, get_options, item_click_callback) {
 
                 switch (e.key) {
                     case "Enter":
-                        if (list_items.length === 0) {
-                            $(e.target).trigger("click");
-                        } else {
-                            // Select first item.
+                        if (e.target === $search_input.get(0)) {
+                            // Select first item if in search input.
                             first_item().trigger("click");
+                        } else if (list_items.length !== 0) {
+                            $(e.target).trigger("click");
                         }
-                        e.target.click();
                         e.stopPropagation();
                         e.preventDefault();
                         break;
```

Reply:
Oops, thanks for fixing!